ipykernel/subshell_manager.py

```diff
@@ -0,0 +1,280 @@
+"""Manager of subshells in a kernel."""
+
+# Copyright (c) IPython Development Team.
+# Distributed under the terms of the Modified BSD License.
+from __future__ import annotations
+
+import typing as t
+import uuid
+from dataclasses import dataclass
+from threading import Lock, current_thread, main_thread
+
+import zmq
+import zmq.asyncio
+from anyio import create_memory_object_stream, create_task_group
+
+from .subshell import SubshellThread
+
+
+@dataclass
+class Subshell:
+    thread: SubshellThread
+    shell_channel_socket: zmq.asyncio.Socket
+
+
+class SubshellManager:
+    """A manager of subshells.
+
+    Controls the lifetimes of subshell threads and their associated ZMQ sockets.
+    Runs mostly in the shell channel thread.
+
+    Care needed with threadsafe access here.  All write access to the cache occurs in
+    the shell channel thread so there is only ever one write access at any one time.
+    Reading of cache information can be performed by other threads, so all reads are
+    protected by a lock so that they are atomic.
+
+    Sending reply messages via the shell_socket is wrapped by another lock to protect
+    against multiple subshells attempting to send at the same time.
+    """
+
+    def __init__(self, context: zmq.asyncio.Context, shell_socket: zmq.asyncio.Socket):
+        assert current_thread() == main_thread()
+
+        self._context: zmq.asyncio.Context = context
+        self._shell_socket = shell_socket
+        self._cache: dict[str, Subshell] = {}
```

Comment:
Self curiosity:

check the difference between attribute typing in `__init__` and at class level.

Is mypy smart enough, or does it say `attribute may be unset` if type set in `__init__` ??